xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -651,163 +1270,174 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
 
-      if (i < scripts.size())
-      {
-        lastScript = scripts[i];
-        lastRunStart = i;
-      }
-      else
-      {
-        break;
-      }
-    }
-  }
+  RunInfo info;
+  info.startOffset = start;
+  info.endOffset = text.size();
+  info.script = scriptLast;
+  info.font = FindFallback(text[info.endOffset - 1].letter, info.fontTable);
+  runs.emplace_back(info);
 
+  hb_buffer_t* buffer = hb_buffer_create();
   for (auto& run : runs)
   {
-    run.m_buffer = hb_buffer_create();
-    hb_buffer_set_direction(run.m_buffer, static_cast<hb_direction_t>(HB_DIRECTION_LTR));
-    hb_buffer_set_script(run.m_buffer, run.m_script);
+    hb_buffer_reset(buffer);
+    hb_buffer_set_direction(buffer, static_cast<hb_direction_t>(HB_DIRECTION_LTR));
+    hb_buffer_set_script(buffer, run.script);
+    hb_buffer_set_content_type(buffer, HB_BUFFER_CONTENT_TYPE_UNICODE);
 
-    for (unsigned int j = run.m_startOffset; j < run.m_endOffset; j++)
+    for (unsigned int j = run.startOffset; j < run.endOffset; j++)
     {
-      hb_buffer_add(run.m_buffer, static_cast<wchar_t>(0xffff & text[j]), j);
+      hb_buffer_add(buffer, text[j].letter, j);
     }
 
-    hb_buffer_set_content_type(run.m_buffer, HB_BUFFER_CONTENT_TYPE_UNICODE);
-    hb_shape(m_hbFont, run.m_buffer, nullptr, 0);
+    if (run.font)
+      hb_shape(run.font->m_hbFont, buffer, nullptr, 0);
+
     unsigned int glyphCount;
-    run.m_glyphInfos = hb_buffer_get_glyph_infos(run.m_buffer, &glyphCount);
-    run.m_glyphPositions = hb_buffer_get_glyph_positions(run.m_buffer, &glyphCount);
+    run.glyphInfos = hb_buffer_get_glyph_infos(buffer, &glyphCount);
+    run.glyphPositions = hb_buffer_get_glyph_positions(buffer, &glyphCount);
 
     for (size_t k = 0; k < glyphCount; k++)
     {
-      glyphs.emplace_back(run.m_glyphInfos[k], run.m_glyphPositions[k]);
+      glyphs.emplace_back(run.glyphInfos[k], run.glyphPositions[k], run.font);
     }
-
-    hb_buffer_destroy(run.m_buffer);
   }
+  hb_buffer_destroy(buffer);
 
   return glyphs;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr, FT_UInt glyphIndex)
+Character* CGUIFontTTF::GetCharacter(const character_t& chr,
+                                     FT_UInt glyphIndex,
+                                     CFontData* fontData)
 {
-  const wchar_t letter = static_cast<wchar_t>(chr & 0xffff);
+  const auto letter = chr.letter;
 
   // ignore linebreaks
-  if (letter == L'\r')
+  if (letter == static_cast<decltype(chr.letter)>('\r'))
     return nullptr;
 
-  const character_t style = (chr & 0x7000000) >> 24; // style = 0 - 6
+  const auto style = chr.style;
+
+  if (!fontData)
+  {
+    glyphIndex = FT_Get_Char_Index(m_fontMain.m_face, chr.letter);
+    if (glyphIndex != 0)
+      fontData = &m_fontMain;
+    else
+    {
+      const CFontTable* fontTable;
+      fontData = FindFallback(chr.letter, fontTable);
+      if (!fontData || (glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter)) != 0)
```

Comment:
I think this should be `== 0` 

`FT_Get_Char_Index(fontData->m_face, chr.letter)) == 0`

`FT_Get_Char_Index` error is when index is zero